neurovault/apps/statmaps/static/scripts/metadata-edit.js

```diff
@@ -0,0 +1,382 @@
+      } else if (datasource.url) {
+        return cachedAjaxSource(datasource.url);
+      }
+    } else {
+      return undefined;
+    }
+  }
+
+  function choiceValidator(choices) {
+    return function(value, callback) {
+      if (!value) {
+        return callback(true);
+      }
+      var elems = NVMetadata.datasources[choices].filter(function(x) {
+        return x === value;
+      });
+      if (elems.length === 0) {
+        return callback(false);
+      }
+      return callback(true);
+    };
+  }
+
+  function getFieldValidator(field) {
+    if (field.required && !field.datasource) {
+      return stringRequiredValidator;
+    } else if (!field.required && field.datasource && field.datasource.choices) {
+      return choiceValidator(field.datasource.choices);
+    } else {
+      return undefined;
+    }
+  }
+
+  function columnSettings(headers) {
+    return headers.map(function (x) {
+      return {
+        type: guessFieldType(x),
+        source: getDataSource(x),
+        strict: x.required,
+        validator: getFieldValidator(x)
+      }
+    });
+  }
+
+  function getDefaultHeight(data) {
+    var minHeight = 400,
+      rowHeight = 24;
+
+    if ((data.length + 1) * rowHeight < minHeight) {
+      return minHeight;
+    } else {
+      return undefined;
+    }
+  }
+
+  function columnExist(columnName) {
+    var value = columnName.trim(),
+      elems = NVMetadata.headers.filter(function (x) {
+        return x.name === value;
+      });
+
+    return elems.length !== 0;
+  }
+
+  function isFixedColumn(index) {
+    return NVMetadata.headers[index].fixed
+  }
+
+  function hasAnyValues(arr) {
+    var i, len;
+    for (i = 0, len = arr.length; i < len; i += 1) {
+      if (!emptyOrSpaced(arr[i])) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function filterEmptyRows(array) {
+    return array.filter(function(x) {
+      return hasAnyValues(x);
+    });
+  }
+
+  function serializeTable(hotInstance) {
+    return [window.NVMetadata.headers.map(function (x) {
+      return x.name
+    })].concat(filterEmptyRows(hotInstance.getData()));
+  }
+
+  function validateInput(value, success, failure) {
+    if (emptyOrSpaced(value)) {
+      failure();
+    } else if (columnExist(value)) {
+      failure("Column already exists");
+    } else {
+      success();
+    }
+  }
+
+  function showError($el, msg) {
+    $el.addClass('has-error');
+    if (msg) {
+      $('.help-block', $el).text(msg);
+    }
+  }
+
+  function clearError($el) {
+    $('.help-block', $el).empty();
+    $el.removeClass('has-error');
+  }
+
+  function getNewColumnName(callback) {
+    var $modalEl = $('#columnNameModal'),
+      $submit = $('button[type="submit"]', $modalEl),
+      $columnNameEl = $('#column-name'),
+      $inputWrapper = $('.js-input-wrapper', $modalEl);
+
+    $modalEl.on('shown.bs.modal', function () {
+      $columnNameEl.focus().attr("autocomplete", "off");
+
+      $columnNameEl.keyup(function (e) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['e' is defined but never used.](https://www.codacy.com/app/krzysztof-gorgolewski/NeuroVault_NeuroVault/file/825092758/issues/source?bid=2767031&fileBranchId=2767194#l217)